### PR TITLE
Import google scholar publications

### PR DESCRIPTION
**Pull request by vgoh1**
# Description 
- Implemented a web scraper to scrape publications off a user's google scholar profile
- Includes validation to check if the scraped publications are already in the db for the team
- Implementing pagination on the frontend + handling edge cases
- Refactoring

# Type of change 
- [ ] Bug fix
- [x] New feature
- [x] Refactoring
- [ ] Documentation 

# Problem
- Wanted to stop using AWS lambda + scholarly library in python to scrape because it's too slow and volatile to being blocked by google (uses selenium)

# Solution description
Uses puppeteer which spawns a headless browser to scrape, which won't be blocked by google if you go on the page too much, also much faster because we don't have to deal with the cold start of AWS lambda
Puppeteer-cluster is used to parallelize the process
Pagination logic is implemented by yours truly and I probably broke so many good react practices in the process so pls review 
Refactored some constants into a separate file so they'd be easy to configure both on frontend and backend

# Tests 
Tested with these profiles: https://scholar.google.com.sg/citations?user=3tyGlPsAAAAJ
https://scholar.google.com.sg/citations?hl=en&user=eRbvWqYAAAAJ

Edge cases: 
If no publications have been selected to import
image

If you reach the end of the user's profile:
image

If all publications from the user have been imported: (logic based on if the most recent publications are in the db or not)
image

# Relevant document/ link (if any) 
I documented my findings using puppeteer at the end of this doc: https://docs.google.com/document/d/1CrZi1HxeSRmjTsOq30PnOb9OFWUmn_djxgjCAYKQZRs/edit

Right now I have it set to returning 10 publications at once, so 10 threads, feel free to discuss or play around with the config if you think it should be another number

# Risk
Medium - important functionality

### CODE REVIEW

**Review by jfan0010 — COMMENTED** [review]
The code LGTM in general! 

Not sure if I'm testing it correctly, but when I'm importing more than 10 publications (tried 20 and 30) on the website using `npm run dev`, it takes quite a while, but no signal shown (as a user idk if they're running in background or not running at all). 

Then I tried to import publications again (while the background is running i guess), and got the error shown in the image.

error message shown when import publications again while background is running

File: client/src/reducers/importedPublicationsReducer.js [hunk 1]
```diff
@@ -1,25 +1,108 @@
...
+  activePage: 0,
+  shownPublications: [],
+  publicationsToImport: []
+};
+
+const toggleActivePage = (state, retrievedPublications) => {
+  if (state.publications.length < pageSize) {
+    return 1
+  } else {
+    if (state.activePage != state.totalPages) { // they weren't on the latest page
```
[the remark this answers is not in the document]
Comment (vgoh1) [reply client/src/reducers/importedPublicationsReducer.js:31]: Will add in next commit

File: client/src/components/publications/importedPublication/ImportedPublication.js [hunk 2]
```diff
@@ -30,27 +35,33 @@ const ImportedPublication = ({pub, index, setChecked}) => {
         setExpand(expand)
```
[the remark this answers is not in the document]
Comment (vgoh1) [reply client/src/components/publications/importedPublication/ImportedPublication.js:35]: Button looks kinda ugly haha 
image
But good point to bring up, I kinda want to show that it's a clickable link like highlight it in blue or something. I'll note it down for next semester

**Comment by vgoh1** [comment]
@jfan0010 I couldn't recreate the error you were getting, could you give me the steps to recreate it?

As a side note I think getting more than 20 publications at once is a bit much unless we make the number of threads = number of publications but I'm not sure what the load would be like. 10 publications/threads takes 5 seconds on average and sometimes 8 seconds on the higher end which I think is okay-ish.

File: client/src/components/publications/importedPublication/ImportedPublication.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (jfan0010) [reply client/src/components/publications/importedPublication/ImportedPublication.js:35]: Actually, I guess we can keep it since it's a third-party website (not a page of Researchify).

Here are some potential solutions I can think of, which might be possible to apply to your case:  
1. Use `href` tag in `<a>`,   and the link will act like a normal link. [Example](https://www.w3schools.com/tags/tag_a.asp)
2. Use Button from React-Bootstrap, with `variant="link"`, to make to button looks like a link. [Example](https://react-bootstrap.github.io/components/buttons/#:~:text=use%20any%20of%20the%20available%20button%20style%20types%20to%20quickly%20create%20a%20styled%20button.%20just%20modify%20the%20variant%20prop.)

But I think we can leave it for now as it wasn't that important  :)

**Comment by jfan0010** [comment]
> @jfan0010 I couldn't recreate the error you were getting, could you give me the steps to recreate it?
> 
> As a side note I think getting more than 20 publications at once is a bit much unless we make the number of threads = number of publications but I'm not sure what the load would be like. 10 publications/threads takes 5 seconds on average and sometimes 8 seconds on the higher end which I think is okay-ish.

@vgoh1 Sure, here're my steps:
1. `npm run dev` in Researchify root folder.
2. (In publication page) Press button `import publication`
3. Enter the link of [Burkard Polster](https://scholar.google.com.sg/citations?hl=en&user=eRbvWqYAAAAJ).
4. Select `show more` at the bottom of the pop up modal.
5. Confirm the selection by pressing `import` (20 of them) -> (The modal is closed)
6. Refresh the page (The 20 publications selected is not successfully imported yet)
7. (The page is loading, no publications shown at the moment) Repeat steps 2 & 3 (while loading).

If we're not letting the user get more than 10 publications per time, I think might as well we can ignore it, and only cap the number of publications users can get in future then.

File: client/src/components/publications/importedPublication/ImportedPublication.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (vgoh1) [reply client/src/components/publications/importedPublication/ImportedPublication.js:35]: I'll go with href tag cause it's nice and simple:
image

Thanks!

**Comment by riordanalfredo** [comment]
Btw, I got an issue as well here:

image

**Review by kdav108 — CHANGES_REQUESTED** [review]
Couldn't complete the entire PR review in one sitting 😂, but so far it looks good. Just two small things I could spot.

File: api/src/config/puppeteer.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  pageSizeSuffix: "&pagesize=", // concatenate page size after
+  sortBySuffix: "&view_op=list_works&sortby=pubdate", // put most recent pubs first
+};
+
+const categoryType = {
+  CONFERENCE: "CONFERENCE",
+  JOURNAL: "JOURNAL",
+  OTHER: "OTHER",
+  BOOK: "BOOK",
+};
```
[the remark this answers is not in the document]
Comment (kdav108) [reply api/src/config/puppeteer.js:20]: Actually don't worry, I don't think we can have a common variable between `client` and `api`

File: api/src/models/publication.model.js [hunk 4]
```diff
@@ -3,63 +3,68 @@
...
-            required: true,
-            minlength: 3
-        },
-        volume: String,
-        issue: String,
-        pages: String,
-        publisher: String
+      type: {
+        type: String,
+        enum: ['CONFERENCE', 'JOURNAL', 'OTHER', 'BOOK'],
```
[the remark this answers is not in the document]
Comment (vgoh1) [reply api/src/models/publication.model.js:50]: Yup about this one, I couldn't figure out how to do it cause mongoose would throw an error about not being able to convert the enum to a string

File: api/src/config/puppeteer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (vgoh1) [reply api/src/config/puppeteer.js:20]: Yeah I couldn't figure that out
